File: question/neetcode150/binary_search/median-of-two-sorted-arrays.py
from typing import List

class Solution:
    def findMedianSortedArrays(self, a1: List[int], a2: List[int]) -> float:
        if len(a1) > len(a2):
            a1,a2 = a2,a1

        n = len(a1)
        m = len(a2)

        i = -1
        l = n
        while l:
            while i+l < n and a1[i+l] <= a2[(n+m)//2-i-l-1]:
                i+=l
            l//=2

        j = (n+m)//2 - (i+1) - 1

        # print(a1, a2, i, j)

        if i == -1:
            mi = a2[j]
        elif j == -1:
            mi = a1[i]
        else:
            mi = max(a1[i],a2[j])
        if i == n-1:
            ma = a2[j+1]
        elif j == m-1:
            ma = a1[i+1]
        else:
            ma = min(a2[j+1],a1[i+1])

        # print(mi, ma)

        if (n+m)%2:
            return ma
        else:
            return (mi+ma)/2

File: question/neetcode150/binary_search/median-of-two-sorted-arrays.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& a1, vector<int>& a2) {
        int n1 = a1.size(), n2 = a2.size();

        if (n1 > n2){
            swap(a1,a2);
            swap(n1,n2);
        }

        int k = -1;
        int l = (n1+n2)/2;
        for(int i=l;i>0;i/=2){
            while(k+i<n1 and a1[k+i] <= a2[l-(k+i+1)]){
                k+=i;
            }
        }

        int a1l = k, a1r = k+1;
        int a2l = l-(k+1)-1, a2r = a2l+1;

        // cout << k << endl;
        // cout << a1l << endl;
        // cout << a1r << endl;
        // cout << a2l << endl;
        // cout << a2r << endl;

        if((n1+n2)%2){
            if(a1r==n1){
                return a2[a2r];
            }
            if(a2r==n2){
                return a1[a1r];
            }
            return min(a1[a1r], a2[a2r]);
        }else{
            int ma, mi;
            if(a1r==n1){
                ma = a2[a2r];
            }
            if(a2r==n2){
                ma = a1[a1r];
            }
            if(a1r!=n1 and a2r!=n2){
                ma = min(a1[a1r], a2[a2r]);
            }

            if(a1l==-1){
                mi = a2[a2l];
            }
            if(a2l==-1){
                mi = a1[a1l];
            }
            if(a1l!=-1 and a2l!=-1){
                mi = max(a1[a1l], a2[a2l]);
            }

            // cout << mi << " " << ma << endl;

            return (mi+ma)/2.0;
        }
    }
};